### PR TITLE
[Merged by Bors] - feat: port Topology.TietzeExtension

### PR DESCRIPTION
**Pull request by apurvnakade**
---

~~I'm getting a `maxHearbeats` error. The file builds if I set it to 250000. See comment below.~~ Resolved.

Please feel free to push to this branch.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/TietzeExtension.lean [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+topological space `Y`. Let `f` be a continuous real-valued function on `X`. Let `t` be a nonempty
+convex set of real numbers (we use `ord_connected` instead of `convex` to automatically deduce this
+argument by typeclass search) such that `f x ∈ t` for all `x`. Then there exists a continuous
+real-valued function `g : C(Y, ℝ)` such that `g y ∈ t` for all `y` and `g ∘ e = f`. -/
+theorem exists_extension_forall_mem_of_closedEmbedding (f : C(X, ℝ)) {t : Set ℝ} {e : X → Y}
+    [hs : OrdConnected t] (hf : ∀ x, f x ∈ t) (hne : t.Nonempty) (he : ClosedEmbedding e) :
+    ∃ g : C(Y, ℝ), (∀ y, g y ∈ t) ∧ g ∘ e = f := by
+  have h : ℝ ≃o Ioo (-1 : ℝ) 1 := orderIsoIooNegOneOne ℝ
+  set F : X →ᵇ ℝ :=
+    { toFun := Subtype.val ∘ h ∘ f
```
Comment (apurvnakade) [inline Mathlib/Topology/TietzeExtension.lean:370]: This was originally `coe ∘ h ∘ f`. I'm guessing that this definition is causing the `masHeartbeats` error.

Comment (apurvnakade) [reply Mathlib/Topology/TietzeExtension.lean:370]: I'm not able to find the direct analogue of this `coe` in `mathlib4`.

Comment (apurvnakade) [reply Mathlib/Topology/TietzeExtension.lean:370]: This is the output of `set_option profiler true`

```lean 
tactic execution of Lean.Parser.Tactic.rewriteSeq took 164ms
typeclass inference of CoeT took 282ms
typeclass inference of CoeT took 277ms
typeclass inference of CoeT took 277ms
typeclass inference of CoeT took 283ms
typeclass inference of CoeT took 282ms
typeclass inference of CoeT took 276ms
typeclass inference of CoeT took 275ms
typeclass inference of CoeT took 281ms
typeclass inference of CoeT took 289ms
typeclass inference of CoeT took 281ms
typeclass inference of CoeT took 285ms
typeclass inference of CoeT took 273ms
typeclass inference of CoeT took 275ms
typeclass inference of CoeT took 277ms
typeclass inference of CoeT took 278ms
typeclass inference of CoeT took 280ms
typeclass inference of CoeT took 276ms
typeclass inference of CoeT took 275ms
typeclass inference of CoeT took 278ms
typeclass inference of CoeT took 277ms
typeclass inference of CoeT took 278ms
typeclass inference of CoeT took 276ms
typeclass inference of CoeT took 274ms
typeclass inference of CoeT took 273ms
typeclass inference of CoeT took 274ms
tactic execution of Lean.Parser.Tactic.rewriteSeq took 230ms
```

Comment (Komyyy) [reply Mathlib/Topology/TietzeExtension.lean:370]: The true cause seems in `set` tactics.
This command replace declared subterms in local hypotheses and goal with the new definition.
Actually, all `set` tactics in this theorem can be replaced with `let`, and it solves timeouts.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**